openscm/units.py

```diff
@@ -0,0 +1,326 @@
+"""
+Unit handling.
+
+Our unit handling library is `Pint <https://github.com/hgrecco/pint>`_. This allows us
+to define units simply as well as providing us with the ability to define contexts.
+
+
+**A note on emissions units**
+
+Emissions are a flux composed of three parts: mass, the stuff being emitted and the
+time period e.g. "t CO2 / yr". As mass and time are part of SI units, all we need to
+define here are emissions units i.e. the stuff. Here we include as many of the
+canonical emissions units, and their conversions, as possible.
+
+For emissions units, there are a few cases to be considered:
+
+- fairly obvious ones e.g. carbon dioxide emissions can be provided in 'C' or 'CO2' and
+  converting between the two is possible
+- less obvious ones e.g. nitrous oxide emissions can be provided in 'N', 'N2O' or
+  'N2ON', we provide conversions
+- case-sensitivity. In order to provide a simplified interface, using all uppercase
+  versions of any unit is also valid e.g. ``unit_registry("HFC4310mee")`` is the same as ``unit_registry("HFC4310MEE")``
+- hyphens and underscores in units. In order to be Pint compatible and to simplify
+  things, we strip all hyphens and underscores from units.
+
+As a convenience, we allow users to combine the mass and the type of emissions to make
+a 'joint unit' e.g. "tCO2" but it should be recognised that this joint unit is a
+derived unit and not a base unit.
+
+By definining these three separate components, it is much easier to track what
+conversions are valid and which are not. For example, as the emissions units are all
+defined as emissions units, and not as atomic masses, we are able to prevent invalid
+conversions. If emissions units were simply atomic masses, it would be possible to
+convert between e.g. C and N2O which would be a problem. Conventions such as allowing
+carbon dioxide emissions to be reported in C or CO2, despite the fact that they are
+fundamentally different chemical species, is a convention which is particular to
+emissions (as far as we can tell).
+
+Finally, contexts are particularly useful for emissions as they facilitate much easier
+metric conversions. With a context, a conversion which wouldn't normally be allowed
+(e.g. tCO2 --> tN2O) is allowed and will use whatever metric conversion is appropriate
+for that context (e.g. AR4GWP100).
+
+Finally, we discuss namespace collisions.
+
+*CH4*
+
+Methane emissions are defined as 'CH4'. In order to prevent inadvertent conversions of
+'CH4' to e.g. 'CO2' via 'C', the conversion 'CH4' <--> 'C' is by default forbidden.
+However, it can be performed within the context 'CH4_conversions' as shown below:
+
+.. code:: python
+
+    >>> from openscm.units import unit_registry
+    >>> CH4 = unit_registry("CH4")
+    >>> CH4.to("C")
+    pint.errors.DimensionalityError: Cannot convert from 'CH4' ([methane]) to 'C' ([carbon])
+
+    # with a context, the conversion becomes legal again
```

Comment:
Just curious, where are people actually expressing CH4 as C?

Reply:
Good question, I can't remember where I've seen it (we can remove if you want but I think we've done the work now so just leave in?)

Reply:
I might use it

Reply:
Fair enough!